dask/dataframe/io/tests/test_sql.py

```diff
@@ -332,3 +335,101 @@ def test_no_character_index_without_divisions(db):
     # attempt to read the sql table with a character index and no divisions
     with pytest.raises(TypeError):
         read_sql_table("test", db, npartitions=2, index_col="name", divisions=None)
+
+
+def test_to_sql():
+    df_by_age = df.set_index("age")
+    df_appended = pd.concat([df, df,])
+
+    # Set this to False to use a sqlite DB file under the current directory, and not delete it (for easier debugging
+    # after tests have run)
+    rm_tmpdir = True
+    if rm_tmpdir:
```

Comment:
Can you use `tempfile` or the `tmpdir` fixture here?

Reply:
I don't think I can… in one case (only enabled if a person wants to debug the tests, and changes the `rm_tmpdir` value), I want to not `rm` the tmpdir. I don't see Dask's custom `tempfile`/`tmpdir` supporting that.

stdlib equivalents do, I believe, but I guessed that the style is to make do with Dask's own versions.

In any case, I could just remove it and always use the intended (`rm_tmpdir == True`) path. This kind of mostly-unreachable test-debugging code is not super common. It was useful when I was iterating on the tests, but maybe not worth the distraction from leaving it in in perpetuity.

Your call / lmk what you'd like.

Reply:
For developing the test this may be fine. But before we merge I think we'd prefer the `tempfile` since it'll be easier to understand the test at a glance.

And I just recalled, rather than `tempfile` we can just use an in-memory sqlite database with `with sqlite3.connect(":memmory:") as con:` (or whatever the sqlalchemy equivalent is if pandas requires that rather than sqlite)

Reply:
The engine connection string for sqlalchemy would be ""sqlite://""

Reply:
conn string "sqlite://" for memory

Reply:
thx, made it just use `tmpfile()`, much cleaner

I tried swapping in `"sqlite://"` as the DB URI, but got a `NoSuchTableError` in one of the tests, with no clear cause (and tests pass w/ `tmpfile()`), so I gave up, but can look at it more if you like.

In-memory version would be nice, but `tmpfile()` mimics what other cases in this file already do.

Reply:
btw, I debugged the in-memory ("`sqlite://`") version some more; it seems that different threads can't hit the same in-memory SQLIte instance, so multi-partition tests fail? `tmpfile()` seems like the way to go.

In the process of debugging, I made `to_sql` and `read_from_sql` generally able to take a `con` (`str` or `Engine`), in Pandas style (as opposed to current `str`-only requirement in both).

[Here's the commit](https://github.com/ryan-williams/dask/commit/842a944d8c22a4e2cc236ef8b658b1299e880eb8); lmk if you'd like added to this PR.

Reply:
> it seems that different threads can't hit the same in-memory SQLIte instance

Unfortunate... I would have expected it to be ok within a process.

>  I made to_sql and read_from_sql generally able to take a con

This is a bad idea, because they will not serialise for sending between processes/machines.

Reply:
> This is a bad idea, because they will not serialise for sending between processes/machines.

I don't think [my implementation](https://github.com/ryan-williams/dask/compare/sql...ryan-williams:in-mem-sqlite) (now 2 commits) has the problem you're worried about.

It uses the passed `Engine` locally (e.g. for loading `meta`), but sends the engine's *URL* to workers, which each establish their own connection.

That may still be problematic for reasons I'm not aware of (are engines stateful in some way that independent connections to the same URI don't inherit?), but I think it gets around the issue you're mentioning (and the tests verify that, to at least some degree).

Reply:
OK, fair enough. Might be nice to have a pickle.dumps(df.dask) test to be certain.